### ISSUE TITLE
Implement BOLT

### ISSUE THREAD

**Issue by equill** [equill/neo4cl#1]
The [Messagepack protocol](http://msgpack.org/index.html) has been [implemented in CL by mbrezu](https://github.com/mbrezu/cl-messagepack), which was the main challenge for implementing BOLT.

Follow the [BOLT spec](https://github.com/neo4j-contrib/boltkit), and use [this blogpost](https://neo4j.com/blog/neo4j-3-0-language-drivers/) for additional hints.

**Comment by equill** [comment]
In progress.

**Comment by equill** [comment]
Making steady progress. So far:

- Connect to server, negotiate Bolt version.
- Authenticate to server.

Next:

- Parse the server's response, and do something useful with the contents.

**Comment by equill** [comment]
New status: we have ignition!

Successfully connecting to the server, negotiating the Bolt version, authenticating, then politely closing the session.

Next actions:

- Clean up the FIXMEs
- Implement `parse-<element>` for the rest of the Packstream types.
    - This will involve deciding how to represent the various Structure types.
    - Currently-handled types are: Null, String and Dictionary. Coincidentally, that's the exact set of types required to handle either a SUCCESS or FAILURE response to a HELLO message.

That ought to clear the way to implement auto-commit transactions, which appears to be the minimum viable driver required to displace the existing HTTP-API wrapper.

**Comment by equill** [comment]
Moving along nicely: it now handles a wider range of server responses and data types.

Parameterised queries and multi-byte characters are among the things known to work, and all 242 tests in Restagraph's internal test-suite pass.

It's not yet complete - most Structure types aren't handled, and it doesn't yet do explicit transactions - but it's almost certainly up to the job of replacing the HTTP API wrapper.

**Comment by equill** [comment]
Found a problem with chunking when the server returns very large quantities of data. I suspect this comes from not properly understanding what happens when the server has more than 65535 bytes to return, and thus has to break the message-body up over multiple chunks.

Discovered this by migrating Webcat to the driver, and searching for Wikipage resources, of which I have a few. All other functionality is working, which is nice. Response is snappy enough, which is also nice, given that I've relegated performance as a non-consideration until all functionality is implemented and working _correctly_.

**Comment by equill** [comment]
Fixed the issue with chunking; mercifully, the solution was simple and straightforward, and worked immediately (yes, of *course* I'm suspicious of that).

I've merged this to `main` and am calling it done for "it works for current requirements" values of done. The Bolt and Packstream implementations are both far from complete, but I'm comfortable with handling all remaining shortcomings as separate-but-related bits of work.
